src/adr/0018-polyglot-strategy.md

```diff
@@ -0,0 +1,158 @@
+# 16. Polyglot codebase
+
+Date: 2020-02-03
+
+## Status
+
+Proposed
+
+## Context & Problem Statement
+
+James is written in Java for a very long time. In recent years, Java modernized a lot after a decade of slow progress.
+
+However, in the meantime, most software relying on the JVM started supporting alternative JVM languages to keep being relevant.
+
+It includes Groovy, Clojure, Scala and more recently Kotlin.
+
+Not being open to those alternative languages can be a problem for James adoption.
+
+## Decision drivers
+
+Nowadays, libraries and framework targeting the JVM are expected to support usage of one or several of these alternative languages.
+
+James being not only a mail server but also a development framework needs to reach those expectations.
+
+At the same time, more and more developers and languages adopt Function Programming (FP) idioms to solve their problems.
+
+## Considered options
+
+1. Let the users figure out how to make polyglot setups
+2. Document the usage of polyglot mailets for some popular languages
+3. Document the usage of polyglot components for some popular languages
+4. Actually implement some mailets in some popular languages
+5. Actually implement some components in some popular languages
+
+About Languages:
+
+I. Clojure
+II. Groovy
+III. Kotlin
+IV. Scala
+
+## Decision
+
+We decided on options 4, 5 and IV.
+
+That means we need to write some mailets in Scala and demonstrate how it's done and then used in a running server.
+
+It also means writing and/or refactoring some server components in Scala, starting where it's the most relevant.
+
+### Positive Consequences 
+
+* Modernize parts of James code
+* Leverage Scala richer FP ecosystem and language to overcome Java limitations on that topic
+* Should attract people that would not like Java
+
+### Negative Consequences 
+
+* Adds even more knowledge requirements to contribute to James
+* Scala build time is longer than Java build time
+
+## Pros and Cons of the Options
+
+### Option 1: Let the users figure out how to make polyglot setups
+
+Pros:
+* We don't have anything new to do
+
+Cons:
+* It's boring, we like new challenges
+* Java is declining despite language modernization, it means in the long term, less and less people will contribute to James
+
+### Option 2: Document the usage of polyglot mailets for some popular languages
+
+Pros:
+* It's not a lot of work and yet it opens James to alternatives and can attract people outside Java developers community
+
+Cons:
+* Documentation without implementation often gets outdated when things move forward
+* We don't really gain knowledge on the polyglot matters as a community and won't be able to help users much
```

Comment:
as I said upper, I guess we already have a contribution matching this option, so maybe we can remove it.

Reply:
It's more a reverse documentation of a past decision. This option was open in the past. It makes sense to document it, don't you think?